File: preparacion PSP/ExamenEj3/cliente/src/main/java/network/AuthInterceptor.java
package network;

import domain.model.CacheAuth;
import jakarta.inject.Inject;
import okhttp3.Credentials;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;

public class AuthInterceptor implements Interceptor {
    private final CacheAuth auth;

    @Inject
    public AuthInterceptor(CacheAuth auth) {
        this.auth = auth;
    }

    public Response intercept(Chain chain) throws IOException {
        Request req = chain.request();
        Request newReq;
        if (auth.getJwt()!=null){
            newReq = req.newBuilder()
                    .header("Authorization", auth.getJwt()).build();
        } else {
            newReq = req;
        }

        Response res = chain.proceed(newReq);
        if (!res.isSuccessful()){
            if (res.code()!=401 && res.code()!=403){
                res.close();
                newReq = req.newBuilder()
                        .header("Authorization", Credentials.basic(auth.getUser(), auth.getPass())).build();
                res = chain.proceed(newReq);
            } else {
                return res;
            }
        }

        if (res.header("Authorization")!=null){
            auth.setJwt(res.header("Authorization"));
        }

        return res;



    }


}

File: Moviles/ExamenJoaquinFernandez2/app/src/main/java/com/example/examen/network/AuthInterceptor.kt
package com.example.examen.network

import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response

class AuthInterceptor :Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val original = chain.request()
        val httpUrl = original.url.newBuilder()
            .build()

        val requestBuilder: Request.Builder = original.newBuilder()
            .url(httpUrl)

        return chain.proceed(requestBuilder.build())
    }
}